cli/src/compile.rs

```diff
@@ -15,6 +15,71 @@ pub fn compile(
     ProjectCompiler::new(print_names, print_sizes).compile(project)
 }
 
+// https://eips.ethereum.org/EIPS/eip-170
+const CONTRACT_SIZE_LIMIT: usize = 24576;
+
+pub struct SizeReport {
+    pub contracts: BTreeMap<String, ContractInfo>,
+}
+
+pub struct ContractInfo {
+    pub size: usize,
+    pub is_test_contract: bool,
+}
+
+impl SizeReport {
+    /// Returns the size of the largest contract, excluding test contracts.
+    pub fn max_size(&self) -> usize {
+        let mut max_size = 0;
+        for contract in self.contracts.values() {
+            if !contract.is_test_contract && contract.size > max_size {
+                max_size = contract.size;
+            }
+        }
+        max_size
+    }
+
+    /// Returns true if all contracts are within the size limit, excluding test contracts.
+    pub fn exceeds_size_limit(&self) -> bool {
+        self.max_size() > CONTRACT_SIZE_LIMIT
+    }
+}
+
+impl Display for SizeReport {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        let mut table = Table::new();
+        table.load_preset(UTF8_FULL).apply_modifier(UTF8_ROUND_CORNERS);
+        table.set_header(vec![
+            Cell::new("Contract").add_attribute(Attribute::Bold).fg(Color::Blue),
+            Cell::new("Size").add_attribute(Attribute::Bold).fg(Color::Blue),
+            Cell::new("Margin").add_attribute(Attribute::Bold).fg(Color::Blue),
+        ]);
+
+        for (name, contract) in &self.contracts {
+            if contract.is_test_contract || contract.size == 0 {
+                continue
+            }
```

Comment:
Done in ae8ca0b. Just curious why that's preferable though—doesn't it result in iterating through the contracts twice? Though I doubt there would be a noticeable performance hit from that

Reply:
`filter` is the same thing as `if !condition {continue}`

Reply:
Right, sorry I meant since `filter` means you now iterate through twice, wheras `if !condition {continue}` means you just iterate once. Though now that I think about the equivalent in JS is `.filter().map()` which also iterates through twice. So I guess the answer is the same in rust, i.e we aren't trying to minimize iterations and prefer the more functional/immutable vars approach 😅